clientapi/routing/room_tagging.go

```diff
@@ -0,0 +1,265 @@
+// Copyright 2019 Sumukha PK
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package routing
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/matrix-org/dendrite/clientapi/auth/authtypes"
+	"github.com/matrix-org/dendrite/clientapi/auth/storage/accounts"
+	"github.com/matrix-org/dendrite/clientapi/httputil"
+	"github.com/matrix-org/dendrite/clientapi/jsonerror"
+	"github.com/matrix-org/dendrite/clientapi/producers"
+	"github.com/matrix-org/gomatrix"
+	"github.com/matrix-org/gomatrixserverlib"
+	"github.com/matrix-org/util"
+)
+
+// newTag creates and returns a new Tag type
+func newTag() gomatrix.TagContent {
+	return gomatrix.TagContent{
+		Tags: make(map[string]gomatrix.TagProperties),
+	}
+}
+
+// GetTags implements GET /_matrix/client/r0/user/{userID}/rooms/{roomID}/tags
+func GetTags(
+	req *http.Request,
+	accountDB *accounts.Database,
+	device *authtypes.Device,
+	userID string,
+	roomID string,
+	syncProducer *producers.SyncAPIProducer,
+) util.JSONResponse {
+
+	if device.UserID != userID {
+		return util.JSONResponse{
+			Code: http.StatusForbidden,
+			JSON: jsonerror.Forbidden("Cannot retrieve another user's tags"),
+		}
+	}
+
+	_, data, err := obtainSavedTags(req, userID, roomID, accountDB)
+
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	dataByte, err := json.Marshal(data)
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	var tagData []gomatrixserverlib.ClientEvent
+	tagContent := newTag()
+	err = json.Unmarshal(dataByte, &tagData)
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	err = json.Unmarshal(tagData[0].Content, &tagContent)
+
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	// send data to syncapi
```

Comment:
This comment doesn't say anything past the code right below it. What data? Why do we send it to sync?

Reply:
Additionally, I wonder if we should have the request wait on this or not. @Cnly should we do this asynchronously do you think?

Reply:
This data is "sent" to `/sync` for it to actually sync data. Based on discussion with @Cnly this block of code was added.

Reply:
Comment will be removed/made more appropriate.

Reply:
Doing this asynchronously sounds sane to me. I'll vote for it.

Reply:
As in `go "task"` ?

Reply:
Generally, yes.

Reply:
I see no async operation here tho... https://github.com/matrix-org/dendrite/blob/master/clientapi/routing/account_data.go#L61

Should all those be changed?

Reply:
Probably, in another PR.